src/optimagic/constraints.py

```diff
@@ -0,0 +1,246 @@
+                "'value' cannot be used with 'lower_bound' or 'upper_bound'."
+            )
+
+        if not isinstance(self.weights, np.ndarray | list | pd.Series | float | int):
+            raise InvalidConstraintError(
+                "'weights' must be an array-like, a pandas Series, a float, or an int."
+            )
+
+        if self.lower_bound is not None and not isinstance(
+            self.lower_bound, float | int
+        ):
+            raise InvalidConstraintError("'lower_bound' must be a float or an int.")
+
+        if self.upper_bound is not None and not isinstance(
+            self.upper_bound, float | int
+        ):
+            raise InvalidConstraintError("'upper_bound' must be a float or an int.")
+
+        if self.value is not None and not isinstance(self.value, float | int):
+            raise InvalidConstraintError("'value' must be a float or an int.")
+
+
+@dataclass(frozen=True)
+class NonlinearConstraint(Constraint):
+    selector: Callable[[PyTree], PyTree] = identity_selector
+    _: KW_ONLY
+    func: Callable[[PyTree], ArrayLike | "pd.Series[float]" | float] | None = None
+    derivative: Callable[[PyTree], PyTree] | None = None
+    lower_bound: ArrayLike | "pd.Series[float]" | float | None = None
+    upper_bound: ArrayLike | "pd.Series[float]" | float | None = None
+    value: ArrayLike | "pd.Series[float]" | float | None = None
+    tol: float | None = None
+
+    def _to_dict(self) -> dict[str, Any]:
+        return {
+            "type": "nonlinear",
+            "selector": self.selector,
+            **_select_non_none(
+                func=self.func,
+                derivative=self.derivative,
+                # In the dict representation, we write _bounds instead of _bound.
+                lower_bounds=self.lower_bound,
+                upper_bounds=self.upper_bound,
+                value=self.value,
+                tol=self.tol,
+            ),
+        }
+
+    def __post_init__(self) -> None:
+        super().__post_init__()
+
+        if _all_none(self.lower_bound, self.upper_bound, self.value):
+            raise InvalidConstraintError(
+                "At least one of 'lower_bound', 'upper_bound', or 'value' must be "
+                "non-None."
+            )
+        if self.value is not None and not _all_none(self.lower_bound, self.upper_bound):
+            raise InvalidConstraintError(
+                "'value' cannot be used with 'lower_bound' or 'upper_bound'."
+            )
+
+        if self.tol is not None and (
+            not isinstance(self.tol, float | int) or self.tol < 0
+        ):
+            raise InvalidConstraintError("'tol' must be non-negative.")
+
+        if self.func is None or not callable(self.func):
+            raise InvalidConstraintError("'func' must be callable.")
+
+        if self.derivative is not None and not callable(self.derivative):
+            raise InvalidConstraintError("'derivative' must be callable.")
+
+
+def pre_process_constraints(
```

Comment:
This function should live in deprecations, so we are aware that it needs to be rewritten at some point.